File: backend/ingredient_allowlist.py
"""Strict allowlist: user ingredients + defined system pantry only."""

from __future__ import annotations

from ingredient_relevance import canonical_ingredient, ingredients_match, normalize_ingredient, parse_user_ingredients
from measurement_units import strip_quantity_prefix

SYSTEM_PANTRY_CANONICAL = {
    "water",
    "salt",
    "black pepper",
    "oil",
    "olive",
    "olive oil",
    "baking powder",
}

FORBIDDEN_UNLESS_USER_HINTS = (
    "berries",
    "cookies",
    "chocolate",
    "fruits",
    "nuts",
    "cream cheese",
    "yogurt",
    "milk",
    "butter",
)


def _user_provided_vanilla(user_ingredients: list[str]) -> bool:
    for item in user_ingredients:
        text = normalize_ingredient(item)
        canon = canonical_ingredient(item)
        if canon == "vanilla" or "vanilla" in text or "וניל" in text:
            return True
    return False


def get_system_pantry_items(user_ingredients: list[str] | None = None) -> list[str]:
    items = ["water", "salt", "black pepper", "oil", "olive oil", "baking powder"]
    if user_ingredients and _user_provided_vanilla(user_ingredients):
        items.extend(["vanilla", "vanilla extract"])
    return items


def is_system_pantry_ingredient(name: str) -> bool:
    canon = canonical_ingredient(strip_quantity_prefix(name or ""))
    if not canon:
        return False
    if canon in SYSTEM_PANTRY_CANONICAL:
        return True
    return any(ingredients_match(name, item) for item in SYSTEM_PANTRY_CANONICAL)


def is_recipe_ingredient_allowed(recipe_line: str, user_ingredients: list[str]) -> bool:
    if not user_ingredients:
        return True

    if any(ingredients_match(user_ing, recipe_line) for user_ing in user_ingredients):
        return True

    return any(
        ingredients_match(pantry_item, recipe_line)
        for pantry_item in get_system_pantry_items(user_ingredients)
    )


def find_unauthorized_recipe_ingredients(recipe: dict, user_ingredients_raw: str) -> list[str]:
    user_ingredients = parse_user_ingredients(user_ingredients_raw)
    if not user_ingredients:
        return []

    unauthorized: list[str] = []
    for item in recipe.get("ingredients") or []:
        if not is_recipe_ingredient_allowed(str(item), user_ingredients):
            unauthorized.append(str(item))
    return unauthorized


def filter_recipe_to_allowed_ingredients(recipe: dict, user_ingredients_raw: str) -> dict:
    user_ingredients = parse_user_ingredients(user_ingredients_raw)
    if not user_ingredients:
        return recipe

    filtered = {
        **recipe,
        "ingredients": [
            item
            for item in (recipe.get("ingredients") or [])
            if is_recipe_ingredient_allowed(str(item), user_ingredients)
        ],
    }
    upgrades = recipe.get("optionalUpgrades") or []
    filtered["optionalUpgrades"] = [
        upgrade
        for upgrade in upgrades
        if is_recipe_ingredient_allowed(
            str(upgrade.get("ingredient", upgrade) if isinstance(upgrade, dict) else upgrade),
            user_ingredients,
        )
    ]
    return filtered

File: src/utils/ingredientAllowlist.js
import { canonicalIngredient, ingredientsMatch, normalizeIngredient } from '../data/ingredientKnowledge'
import { parseUserIngredients } from './ingredientRelevance'
import { stripQuantityPrefix } from './measurementUnits'
import {
  DESSERT_BASIC_PANTRY_CANONICAL,
  isBasicPantryMarkedLine,
  stripBasicPantryLabel,
} from './dessertRecipeBuilder'

/** System pantry items always allowed when the user listed ingredients. */
export const SYSTEM_PANTRY_CANONICAL = new Set([
  'water',
  'salt',
  'black pepper',
  'pepper',
  'oil',
  'olive',
  'olive oil',
  'garlic',
  'onion',
  'baking powder',
  'paprika',
  'cumin',
  'oregano',
  'basil',
  'thyme',
  'cinnamon',
  'nutmeg',
  'ginger',
  'carrot',
  'lentils',
  'lentil',
])

/** Highlights for prompts — not allowed unless the user provided them. */
export const FORBIDDEN_UNLESS_USER_HINTS = [
  'berries',
  'cookies',
  'chocolate',
  'fruits',
  'nuts',
  'cream cheese',
  'yogurt',
  'milk',
  'butter',
]

function userProvidedVanilla(userIngredients) {
  return userIngredients.some((item) => {
    const text = normalizeIngredient(item)
    const canon = canonicalIngredient(item)
    return canon === 'vanilla' || text.includes('vanilla') || text.includes('וניל')
  })
}

export function getSystemPantryItems(userIngredients = []) {
  const items = [
    'water',
    'salt',
    'מלח',
    'black pepper',
    'פלפל שחור',
    'oil',
    'olive oil',
    'שמן זית',
    'שמן',
    'garlic',
    'שום',
    'onion',
    'בצל',
    'baking powder',
  ]
  if (userProvidedVanilla(userIngredients)) {
    items.push('vanilla', 'vanilla extract')
  }
  return items
}

export function isSystemPantryIngredient(name) {
  const canon = canonicalIngredient(stripQuantityPrefix(String(name ?? '')))
  if (!canon) return false
  if (SYSTEM_PANTRY_CANONICAL.has(canon)) return true
  return ['water', 'salt', 'black pepper', 'oil', 'olive oil', 'baking powder'].some((item) =>
    ingredientsMatch(name, item),
  )
}

/**
 * True when a recipe ingredient is allowed for the user's available set.
 * Preference mode (no user ingredients) allows any ingredient.
 */
export function isRecipeIngredientAllowed(recipeLine, userIngredients) {
  if (!userIngredients?.length) return true

  if (userIngredients.some((userIng) => ingredientsMatch(userIng, recipeLine))) {
    return true
  }

  if (isBasicPantryMarkedLine(recipeLine)) {
    const base = stripBasicPantryLabel(stripQuantityPrefix(String(recipeLine ?? '')))
    const canon = canonicalIngredient(base)
    if (canon && DESSERT_BASIC_PANTRY_CANONICAL.has(canon)) return true
    return [...DESSERT_BASIC_PANTRY_CANONICAL].some((item) => ingredientsMatch(base, item))
  }

  return getSystemPantryItems(userIngredients).some((pantryItem) =>
    ingredientsMatch(pantryItem, recipeLine),
  )
}

export function findUnauthorizedRecipeIngredients(recipe, userIngredientsRaw) {
  const userIngredients = parseUserIngredients(userIngredientsRaw)
  if (!userIngredients.length) return []

  const unauthorized = []
  for (const item of recipe.ingredients ?? []) {
    if (!isRecipeIngredientAllowed(item, userIngredients)) {
      unauthorized.push(item)
    }
  }
  return unauthorized
}

export function filterRecipeToAllowedIngredients(recipe, userIngredientsRaw) {
  const userIngredients = parseUserIngredients(userIngredientsRaw)
  if (!userIngredients.length) return recipe

  return {
    ...recipe,
    ingredients: (recipe.ingredients ?? []).filter((item) =>
      isRecipeIngredientAllowed(item, userIngredients),
    ),
    optionalUpgrades: (recipe.optionalUpgrades ?? []).filter((upgrade) =>
      isRecipeIngredientAllowed(upgrade?.ingredient ?? upgrade, userIngredients),
    ),
  }
}
